### ISSUE TITLE
Endless loop in client upon close() when using wss

### ISSUE THREAD

**Issue by downtownallday** [TooTallNate/Java-WebSocket#171]
First, thanks so much for creating the library, it's really useful. I am using this via gottox/socket-io (client) and have run into this nasty problem of an endless loop in SSLSocketChannel2.java when closing a connection (maybe after the other side closed first). A couple of these failures is all it takes to pin the CPU at 100% as the threads that execute these loops pile up.

I doubt this is the correct fix, but what I did was change the following in SSLSocketChannel2.readRemaining(), which seems to work.

```java
		if( inCrypt.hasRemaining() ) {
			unwrap();
			int amount = transfereTo( inData, dst );
>>>			if (engineStatus == SSLEngineResult.Status.CLOSED) {
>>>			    return -1;
>>>			}
			if( amount > 0 )
				return amount;
		}
```
EDIT by @Davidisudadi: formatted code

**Comment by xiaohuca** [comment]
Hi @Davidiusdadi, could you please merge the code change by @downtownallday and make a tag out of it? I am working on a project using https://github.com/Gottox/socket.io-java-client that depends on this library. I am having this high CPU usage issue due to this issue. Thanks.

**Comment by srikanthhugar** [comment]
Hello,

    We are also using socket.io-java-client  and it depends on websocket and noticed CPU and memory is getting consumed like anything.

Is it a same issue? How can i take the changes?

When i analysed the issue from heap dump, noticed below mentioned instances and memory is huge.

class sun.security.ssl.EngineArgs 
class int[] 
class javax.net.ssl.SSLEngineResult 
class java.nio.ByteBuffer[]

**Comment by BrushfireDigitalServices** [comment]
#Update:

- The missing shutdown statement was **fixed** on the master branch in 1c43128.
- Pull request #276 **is now merged** with the master branch as of ea7e061.
- The current master branch of this library is **the most up-to-date version**.
- There is **not yet** test coverage as per #318 to ensure this issue does not resurface. 
- No one is currently assigned to writing the tests. If needed I will write them, though ETA would be late next week at the earliest. **Does someone in the community want to provide tests for these fixes?**
    -The changes to WebSocketClient.onWebsocketClose proposed by @sghosh326 **are not included**.

To everyone who worked hard for the fixes to be included in the main project,  **thank you for your interest and dedication!** Apologies on behalf of the entire Java-Websockets project that this issue was not resolved sooner.

To take the changes, pull commit ea7e061 or later. As discussed in #286, the jar is not up to date. I am closing this issue in hopes that it is now fixed for everyone. If CPU usage or memory leak problems persist, please let the community know as soon as possible.
